documentation/approximate-nn-hnsw.html

```diff
@@ -0,0 +1,98 @@
+    }
+  }
+  index {
+    hnsw {
+      max-links-per-node: 24
+      neighbors-to-explore-at-insert: 500
+    }
+  }
+ }
+</pre>
+<p>In the example above we enable fast approximate search by building an hnsw index over the image_sift_encoding tensor
+
+<ul>
+<li><b>max-links-per-node</b> Specifies how many links per HNSW node to select when building the graph. In <a href="https://github.com/nmslib/hnswlib/blob/master/ALGO_PARAMS.md">HNSWlib</a>
+this parameter is known as <em>M</em> </li>
+<li><b>neighbors-to-explore-at-insert</b> Specifies how many neighbors to explore when inserting a document in the HNSW graph. In HNSWlib this parameter is called <em>ef_construction</em></li>
+</ul>
+
+<p>Choosing the value of these parameters affects both accuracy, query speed and also indexing performance. </p>
+
+<h3 id="approximate-nearest-neighbor-query">Using approximate (fast) nearest neighbor search in query</h3>
+<p>With a hnsw index enabled on the tensor we can chose if we want to use approximate or exact matching by the <em>approximate</em> annotation:</p>
+
+<pre>
+{
+  "yql": "select title,price from sources * where ([{\"targetNumHits\": 10, \"approximate\":false}]nearestNeighbor(image_sift_encoding,query_vector_sift)) and in_stock = true;",
+  "hits": 100
+  "ranking.features.query(query_vector_sift): [0.21,0.12,....],
+  "ranking.profile": "image_similarity" 
+}
+</pre>
+<p>By default, <em>approximate</em> is true when searching a tensor with a <em>hnsw</em> index. 
+The <em>approximate</em> parameter allows us to compute the accuracy loss by having the ability to compute the exact neighbors and compare with
+the approximation to compute the recall@k. In addition to the K (<em>targetNumHits</em>) there is a <em>hnsw.exploreAdditionalHits</em> parameter 
+which controls how many extra nodes in the graph (in addition to <em>targetNumHits</em>) should be explored before selecting the best closest k hits. 
+Using a greater number gives better quality (accuracy as compared with the brute force) but worse performance.
+
+<pre>
+{
+  "yql": "select title,price from sources * where ([{\"targetNumHits\": 10, \"hnsw.\":false}]nearestNeighbor(image_sift_encoding,query_vector_sift)) and in_stock = true;
+  "hits": 100
+  "ranking.features.query(query_vector_sift): [0.21,0.12,....],
+  "ranking.profile": "image_similarity" 
+}
+</pre>
+
+
+<h2 id="nearest-neighbor-search-considerations">Nearest Neighbor Search Considerations</h2>
+<ul>
+<li>When using the <em>nearestNeighbor</em> operator the total hit count is not accurate. This is similar to the behavior with weakAnd. This is true for both approximate and brute force.</li>
+<li>When using the <em>nearestNeighbor</em> operator the <a href="grouping.html">grouping</a> count is not accurate. </li>
+<li>Changing the <em>distance-metric</em> for tensor with hnsw index requires re-indexing, similar changing the <em>max-links-per-node</em> and <em>neighbors-to-explore-at-insert</em> requires re-indexing.</li>
```

Comment:
Changing the index settings only requires re-starting: https://docs.vespa.ai/documentation/reference/schema-reference.html#changes-that-require-restart-but-not-re-feed.

Reply:
what about distance?

Reply:
That is also handled with re-starting.